core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceCreator.scala

```diff
@@ -0,0 +1,688 @@
+    * thrown.  If valid, then return the Union of sourceWorkspaceADs and destWorkspaceADs.
+    *
+    * TODO: If sourceWorkspaceADs is a subset of destWorkspaceADs, isn't the Union equivalent to destWorkspaceADs?
+    *
+    * @param sourceWorkspaceADs
+    * @param destWorkspaceADs
+    * @return
+    */
+  private def unionAuthDomains(sourceWorkspaceADs: Set[ManagedGroupRef], destWorkspaceADs: Set[ManagedGroupRef]): Set[ManagedGroupRef] = {
+    // if the source has an auth domain, the dest must also have that auth domain as a subset
+    // otherwise, the caller may choose to add to the auth domain
+    if (sourceWorkspaceADs.subsetOf(destWorkspaceADs)) sourceWorkspaceADs ++ destWorkspaceADs
+    else {
+      val missingGroups = sourceWorkspaceADs -- destWorkspaceADs
+      val errorMsg = s"Source workspace has an Authorization Domain containing the groups ${missingGroups.map(_.membersGroupName.value).mkString(", ")}, which are missing on the destination workspace"
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.UnprocessableEntity, errorMsg))
+    }
+  }
+
+  /**
+    * Copy all of the Entities AND all of the Method Configs from sourceWorkspace into destWorkspace.
+    *
+    * This was "All or nothing" before, so keeping it that way.  If either Entities OR Method Configs fail to copy over,
+    * then neither will copy over.
+    *
+    * @param sourceWorkspace
+    * @param destWorkspace
+    * @return
+    */
+  private def copyEntitiesAndMethodConfigs(sourceWorkspace: Workspace, destWorkspace: Workspace): Future[Unit] = {
+    dataSource.inTransaction { dataAccess =>
+      dataAccess.entityQuery.copyAllEntities(sourceWorkspace, destWorkspace) andThen
+        dataAccess.methodConfigurationQuery.listActive(sourceWorkspace).flatMap { methodConfigShorts =>
+          val inserts = methodConfigShorts.map { methodConfigShort =>
+            dataAccess.methodConfigurationQuery.get(sourceWorkspace, methodConfigShort.namespace, methodConfigShort.name).flatMap { methodConfig =>
+              dataAccess.methodConfigurationQuery.create(destWorkspace, methodConfig.get)
+            }
+          }
+          DBIO.seq(inserts: _*)
+        } andThen {
+        DBIO.successful()
+      }
+    }
+  }
+
+  /**
+    * Find all the objects in sourceWorkspace's bucket that start with filePrefix and copy them to destWorkspace's
+    * bucket
+    *
+    * @param filePrefix
+    * @param sourceWorkspace
+    * @param destWorkspace
+    */
+  private def copyBucketContents(sourceWorkspace: Workspace, destWorkspace: Workspace, filePrefix: Option[String]): Future[List[Option[StorageObject]]] = {
+    filePrefix match {
+      case Some(prefix) => gcsDAO.listObjectsWithPrefix(sourceWorkspace.bucketName, prefix).flatMap { objectsToCopy =>
+        Future.traverse(objectsToCopy) { objectToCopy => gcsDAO.copyFile(sourceWorkspace.bucketName, objectToCopy.getName, destWorkspace.bucketName, objectToCopy.getName) }
```

Comment:
probably a weakness in the original code: doing `Future.traverse` on a possibly large list has the potential to take down a rawls instance. Depending on the execution context it can a) block other work from happening by hogging the queue or b) exhaust memory by causing too many threads to be created.

Reply:
Sure, I didn't mess around in this block of code, just copied it over.  I'm inclined to just let it be for now so we can get through the bulk of this refactor, but if you want to dig into this and try to fix it now, just say the words.